### PR TITLE
Replace FOREACH_WORD_QUOTED with a loop using unquote_first_word in config_parse_exec()

### PR DESCRIPTION
**Pull request by filbranden**
@mbiebl @zonque @mezcalero 

I'm fairly happy with this patch set and I'm fairly confident it fixes most of the recent quoting/escaping regressions.

The commit description indicates differences from previous versions, for instance handling of ; and \; is more strict now, please confirm you're OK with those changes.

Cheers,
Filipe

### CODE REVIEW

File: src/core/load-fragment.c [hunk 1]
```diff
@@ -532,156 +532,149 @@ int config_parse_exec(
...
                 bool separate_argv0 = false, ignore = false;
 
-                path = NULL;
-                nce = NULL;
-                n = NULL;
+                semicolon = false;
 
-                rvalue += strspn(rvalue, WHITESPACE);
+                r = unquote_first_word(&p, &firstword, UNQUOTE_CUNESCAPE|UNQUOTE_UNESCAPE_RELAX);
+                if (r < 0)
```
Comment (filbranden) [inline src/core/load-fragment.c]: Do I need a `log_*_errno` here? Other places in this function will log, and I also have `return log_oom()` elsewhere... What would be an appropriate log for places where I'm returning `-EINVAL` or similar errors from `unquote_first_word`?

File: src/core/load-fragment.c [hunk 2]
```diff
@@ -532,156 +532,149 @@ int config_parse_exec(
...
+                path = NULL;
+                nce = NULL;
 
-fail:
-        n[k] = NULL;
-        strv_free(n);
-        free(path);
-        free(nce);
+                rvalue = p;
+        } while (semicolon);
```
Comment (filbranden) [inline src/core/load-fragment.c:680]: An alternative here is `} while (*p);` and get rid of the `semicolon` flag variable... I haven't tested that fully though. Any preferences?

File: src/shared/util.c [hunk 3]
```diff
@@ -5292,33 +5234,49 @@ int unquote_first_word(const char **p, char **ret, UnquoteFlags flags) {
...
+                                        s[sz++] = c; /* normal explicit char */
                                 else
-                                        sz += utf8_encode_unichar(s + sz, u);
+                                        sz += utf8_encode_unichar(s + sz, u); /* unicode chars we'll encode as utf8 */
                         } else
                                 s[sz++] = c;
 
-                        state = DOUBLE_QUOTE;
+end_escape:
+                        state = (state == SINGLE_QUOTE_ESCAPE) ? SINGLE_QUOTE :
```
Comment (filbranden) [inline src/shared/util.c]: An alternative here is to keep `state` and `escape` separate, perhaps with a bool escape.

Or turn `ESCAPE` into a bit value like 0x10 or 0x20, etc. and use `SINGLE_QUOTE | ESCAPE`, etc. so that we could simply do `state &= ~ESCAPE` here to go back to the original state.

Any preferences either way?

File: src/core/load-fragment.c (hunk 1 above)
Comment (zonque) [inline src/core/load-fragment.c]: Yes, I think that would help users to spot problems in their unit files. Just use log_syntax(), as that will already include the current line and file that is parsed, and augment that with whatever caused the error (in this case, 'p').

Comment (filbranden) [inline src/core/load-fragment.c]: Done.

Using `log_syntax()` for errors from `unquote_first_word()` and using `log_oom()` if some `strv` operation fails (this is a pattern I've also seen elsewhere.)

**Comment by filbranden** [comment]
I updated the commit to:
- Rebase to latest sources, including upstream fix for escaping of semicolon as `\073`
- Added additional test cases, especially around handling of semicolon, both on its own, escaped and quoted.
- Added logging to all code paths that return an error.
- Kept original behavior of returning 0 instead of -EINVAL on failure of unquoting (e.g. unbalanced quotes, etc.)

I think this one is ready to go. Would you please take another look? Let me know if there's anything that needs to be addressed (or feel free to take over, download the pull request and do small changes before you push it.)

Cheers,
Filipe

**Comment by filbranden** [comment]
cc @arvidjaar who commented on this patchset before.

**Comment by filbranden** [comment]
Thanks for your comments!

I'd appreciate it if you could answer my questions above. I'll try to send you an updated patchset over the weekend.

Cheers,
Filipe

**Comment by filbranden** [comment]
Updated according to the comments and added a lot more tests.

Also split the new tests into a separate prequel commit, so that the behavior changes would be very explicit in the changes to the test results.

Tested it to make sure `make check` is still working as expected and also installed and booted with it, didn't see any problems.

Please take another look.

Cheers,
Filipe

**Comment by poettering** [comment]
Unfortunately it doesn't apply anymore! Sorry for the long delay in reviewing. Any chance you can rebase this? Will merge then!

**Comment by filbranden** [comment]
I'll rebase and retest today...

Or if you prefer to delay this until after v221 release just let me know and I'll postpone it.

Ok to reuse the same PR? I guess as there are no line comments really then it should be fine...

If I don't hear from you again about postponing for after v221, then expect rebased and retested later today from me.

Cheers,
Filipe

**Comment by poettering** [comment]
sounds good to merge this before the release. it has tests and stuff after all. reusing the PR is fine, though usually we rty to avoid it...

**Comment by martinpitt** [comment]
I don't want to make a big fuss, but we've had way more than just one regression due to parser changes in the past, including really bad ones which crashed pid 1. Usually stuff like that isn't the best thing to land two days before making a new release..

**Comment by filbranden** [comment]
@martinpitt Yeah I kind of agree with you... I'll leave the final decision to @poettering though, so I think I'll rebase and push and then this can be merged either before or after v221 (I don't expect it to get out of sync between now and then.)

If you prefer to only merge parts of it now (such as the extra unit tests) I'd be happy to move only those commits that do not change behavior to another PR so that it can be merged now (or feel free to just check out this PR branch and cherry-pick them yourself.)

Cheers!
Filipe

**Comment by poettering** [comment]
There are two days left, and it's the days where people probably test the most, hence please go ahead with the full patch set.

**Comment by filbranden** [comment]
Rebased (no conflicts, git did all the job by itself), retested with unit tests, also installed it and booted my machine with it, looked at output of `journalctl` and `systemctl` and everything looks sane.

So I'd say it's good to go.

Cheers,
Filipe

**Comment by filbranden** [comment]
Awesome thanks!

@martinpitt Let me know if you find any regressions, hopefully the test coverage is enough to prove this works as supposed to and the minor changes in behavior of corner cases shouldn't really affect any real use cases.

**Comment by martinpitt** [comment]
FTR, I turned the crank on my "build packages from upstream trunk" machinery again, and it passed both our integration tests as well as a manual one on a standard desktop install. We don't use any "fancy" Exec\* statements there, but at least any potential regression shouldn't be so bad that we break the boot or anything. Sorry, I'm maybe just being a bit over-cautious after the previous trouble :-)

**Comment by filbranden** [comment]
@martinpitt Totally understandable.

On the other hand, at one point or another this will end up hitting distros and that's when we're likely to find any problems, so whether it would be v221 or v222 doesn't seem to matter that much... Probably better to push it earlier since then we figure out if there's a problem earlier and deal with it, possibly in time to provide distros with a nice backport to fix it downstream.

Cheers!
Filipe

**Comment by poettering** [comment]
@martinpitt thanks a lot! of course best would be to make that machinery run continously! the more CI the merrier! And we'll promise from upstream to look at it all the time.

**Comment by martinpitt** [comment]
@poettering: absolutely. Right now I run it on my laptop daily, and upload the generated packages to https://launchpad.net/~pitti/+archive/ubuntu/systemd (this will only get packages whose tests succeeded). So other users can easily try trunk.

This surely ought to run in some cloud instances, I'm working on that. But this might still be a few weeks, this is much more demanding than a `make check` as the integration tests wreck the testbed in various ways, reboot it dozens of times, etc. We do that with distro packages, but we don't yet have the machinery to run it on PPAs.
